src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -124,61 +129,95 @@ def setUp(self):
         self.d_log = DaosLog(self.context)
 
     def tearDown(self):
+        """Tear down after each test case."""
         super(TestWithoutServers, self).tearDown()
 
         if self.fault_file:
             os.remove(self.fault_file)
 
+
 class TestWithServers(TestWithoutServers):
-    '''
-    Run tests with DAOS servers
+    """Run tests with DAOS servers and at least one client.
+
+    Optionally run DAOS clients on specified hosts.  By default run a single
+    DAOS client on the host executing the test.
 
     :avocado: recursive
-    '''
+    """
+
     def __init__(self, *args, **kwargs):
+        """Initialize a TestWithServers object."""
         super(TestWithServers, self).__init__(*args, **kwargs)
 
         self.agent_sessions = None
 
     def setUp(self):
+        """Set up each test case."""
         super(TestWithServers, self).setUp()
 
-        self.server_group = self.params.get("server_group", '/server/',
-                                            'daos_server')
-        self.hostlist_servers = self.params.get("test_machines", '/run/hosts/*')
-
+        self.server_group = self.params.get(
+            "server_group", "/server/", "daos_server")
+
+        # Determine which hosts to use as servers and optionally clients.
+        # Support the use of a host type count to test with subsets of the
+        # specified hosts lists
+        test_machines = self.params.get("test_machines", "/run/hosts/*")
+        test_servers = self.params.get("test_servers", "/run/hosts/*")
+        test_clients = self.params.get("test_clients", "/run/hosts/*")
+        server_count = self.params.get("server_count", "/run/hosts/*")
+        client_count = self.params.get("client_count", "/run/hosts/*")
```

Comment:
Not all existing `.yaml` files conform to this format though do they?  If that's true, don't they need to be updated to conform to this standard in this patch?

And don't existing tests that fetch node list keys need to have that removed so that these changes are being effected for those tests rather than them then redoing these operations?

Reply:
The test's yaml files would only need to be updated if the test class inherited `TestWithServers`.  to be honest I haven't made it through all of #444 to see if there are conflicts. 

There is some disparity in the server and client yaml specification implementations across the functional tests.  Only includes tests that specifying servers and clients:  
- 3 yaml files use/run/hosts/test_[servers|clients]
- 3 yaml files use /run/hosts/test_machines/test_[servers|clients]
(The above will work with this patch; the following require minor changes)
- 1 yaml file uses /run/hosts/test_servers + /run/hosts/clients
- 2 yaml files use /run/hosts/[servers|clients]
- 1 yaml file uses /run/hosts/test_machines/daos_servers

If those tests are moved to inheriting `TestWithServers` then as part of that code change:
- the yaml files should be updated to at least specify hosts with `test_servers` or `test_clients` under `/run/hosts` or `test_machines` under `/run/hosts`.
- the test-specific implementation of `setUp()` should leverage as much from `TestWithServers.setUp()` as possible, which should include obtaining the server/client lists - to avoid unnecessary code.

Reply:
Yeah.  I'm really hoping #444 can be landed before it conflicts with master again.  I've already had to spend a considerable amount of time rebasing due to the many-files-with-repeated-code problem and another landing that happened just previously.

Before we land any more patches that touch a lot of tests, we need to land #444 so as to reduce the amount of work those patches will have/cause.

Reply:
I've merged the pull request with the #444 changes.

Reply:
I detect 21 tests that currently inherit TestWithServers.  Of those 21 test's yaml files, 20 of them use "test_machines" to define their list of servers.  They will work without modification with this code.  The container/open_close.py test defines its servers via "test_servers" in its yaml file.  This will also not require a modification to run with this implementation of TestWithServers.

Detected test yaml list:
/home/pahender/daos/src/tests/ftest/object/array_obj_test.yaml
/home/pahender/daos/src/tests/ftest/object/obj_update_bad_param.yaml
/home/pahender/daos/src/tests/ftest/object/punch_test.yaml
/home/pahender/daos/src/tests/ftest/object/obj_open_bad_param.yaml
/home/pahender/daos/src/tests/ftest/object/obj_fetch_bad_param.yaml
/home/pahender/daos/src/tests/ftest/container/delete.yaml
/home/pahender/daos/src/tests/ftest/container/create.yaml
/home/pahender/daos/src/tests/ftest/container/global_handle.yaml
/home/pahender/daos/src/tests/ftest/container/attribute.yaml
/home/pahender/daos/src/tests/ftest/container/full_pool_container_create.yaml
/home/pahender/daos/src/tests/ftest/container/container_async.yaml
/home/pahender/daos/src/tests/ftest/container/open.yaml
/home/pahender/daos/src/tests/ftest/container/basic_tx_test.yaml
/home/pahender/daos/src/tests/ftest/pool/global_handle.yaml
/home/pahender/daos/src/tests/ftest/pool/connect_test.yaml
/home/pahender/daos/src/tests/ftest/pool/pool_svc.yaml
/home/pahender/daos/src/tests/ftest/pool/rebuild_no_cap.yaml
/home/pahender/daos/src/tests/ftest/pool/simple_create_delete_test.yaml
/home/pahender/daos/src/tests/ftest/pool/evict_test.yaml
/home/pahender/daos/src/tests/ftest/daos_test/daos_core_test.yaml

Reply:
I see.  Fair enough then.